applications/finetune-quora-embeddings/benchmark.py

```diff
@@ -0,0 +1,354 @@
+                yield batch
+                batch = []
+
+    if batch:
+        yield batch
+
+
+def extract_unique_sentences(test_dataset: Dataset, sentence_to_id_mapping: dict):
+    sentence_pair_elements_1 = []
+    sentence_pair_elements_2 = []
+    labels = []
+
+    for row in test_dataset:
+        s1, s2 = row["questions"]["text"]
+        sentence_pair_elements_1.append(sentence_to_id_mapping[s1])
+        sentence_pair_elements_2.append(sentence_to_id_mapping[s2])
+        labels.append(1 if row["is_duplicate"] else 0)
+
+    return sentence_pair_elements_1, sentence_pair_elements_2, labels
+
+
+@stub.function(image=image, volumes={DATASET_DIR: DATASET_VOLUME})
+def generate_dataset_split():
+    from datasets import load_dataset, load_from_disk
+
+    dataset_path = f"{DATASET_DIR}/{DATASET_NAME}"
+
+    if os.path.exists(dataset_path):
+        dataset = load_from_disk(f"{DATASET_DIR}/{DATASET_NAME}")
+        return
+
+    dataset = load_dataset(DATASET_NAME)
+
+    dataset.save_to_disk(dataset_path)
+    DATASET_VOLUME.commit()
+
+
+def generate_quora_input_example(examples):
+    from sentence_transformers import InputExample
+
+    return [
+        InputExample(
+            texts=[
+                example["questions"]["text"][0],
+                example["questions"]["text"][1],
+            ],
+            label=int(example["is_duplicate"]),
+        )
+        for example in examples
+    ]
+
+
+@stub.function(
+    image=image, volumes={DATASET_DIR: DATASET_VOLUME}, gpu=GPU_CONFIG, timeout=1200
+)
+async def benchmark_mteb_model(model_name):
+    from datasets import load_from_disk
+    from sentence_transformers import util, SentenceTransformer
+    import numpy as np
+    import torch.nn as nn
+    import torch
+
+    dataset = load_from_disk(f"{DATASET_DIR}/{DATASET_NAME}")
+    test_dataset = dataset["test"]
+    model = SentenceTransformer(model_name)
+
+    sentence_to_embedding_map = dict()
+    batch_size = 5000
+    sentences = get_unique_sentences(
+        test_dataset, sentence_to_embedding_map, batch_size
+    )
+    embeddings = []
+    for item in sentences:
+        embeddings.extend(model.encode(item))
+
+    embeddings_tensor = torch.tensor(np.array(embeddings)).to("cuda")
+    # Create an embedding layer with pre-trained weights
+    embedding_layer = nn.Embedding.from_pretrained(embeddings_tensor)
+
+    sentence_pairs_elem_1, sentence_pairs_elem_2, labels = extract_unique_sentences(
+        test_dataset, sentence_to_embedding_map
+    )
+    sentence_pairs_embedding_1 = embedding_layer(
+        torch.as_tensor(sentence_pairs_elem_1, device="cuda")
+    )
+    sentence_pairs_embedding_2 = embedding_layer(
+        torch.as_tensor(sentence_pairs_elem_2, device="cuda")
+    )
+
+    cosine_scores = util.cos_sim(sentence_pairs_embedding_1, sentence_pairs_embedding_2)
+    predictions = np.diag(cosine_scores.cpu()).tolist()
+    return {name: f(labels, predictions) for name, f in EVALS.items()}
```

Comment:
if you also track `avg_cosine = np.mean(predictions)` You'll realise the mean is about .76
which does not even touch 0.3, 0.5, 0.7 at all